Implement long option (#18) and fix small bugs

DevManu-de opened the pull request:
Hey,
I have found that a `(null)` is printed when a unknown option is given.
This pull request fixes this issiue by moving a `argv0` some lines up.

Thanks for the info of my misconfigured git, I think I fixed it now.

Regards,
Manuel

DevManu-de committed:
Fixed a bug where (null) is printed with unknown option

DevManu-de committed:
Reformatted code

DevManu-de committed:
Added global variable to avoid unnecessary arguments

DevManu-de committed:
Fixed a variable

DevManu-de commented:
I have added 2 new files args.h file and args.c.
The args.c collects all arguments and sets the global variable which can be used in oder parts of psh with `extern <data type> <var name>` without the need of any import.

Maybe we should rename the global variable to avoid multiple variables with the same name later.

myzhang1029 committed:
Change function name and arguments

setargs doesn't seem to be very intuitive
and supplying optstring to it doesn't seem to make much sense, since
`optstring` defined what it should do.

myzhang1029 commented:
> I have added 2 new files args.h file and args.c.
> The args.c collects all arguments and sets the global variable which can be used in oder parts of psh with `extern <data type> <var name>` without the need of any import.
Yes, that is great!

> Maybe we should rename the global variable to avoid multiple variables with the same name later.
 True. I really like Lua's way of putting everything into a `internal_state` structure, and that is being worked on in #26.

Also, this project is licensed under GNU GPLv3, as a small project, we do not have a CLA to sign. However, if you are comfortable with licensing your code under that term, I believe it is better if you could add a license note (just like those in other files) with your name to `args.c`, `args.h`, and other files you have contributed to. And maybe we should start signing-off commits now.

DevManu-de committed:
Added license info to args.c & args.h

DevManu-de committed:
Changed verbose variable to VerbosE to avoid duplication

DevManu-de committed:
Added --version flag

DevManu-de committed:
Added --verbose and fixed a bug with -- arguments being collected by the default case

DevManu-de commented:
To comply with the license I added the license info to args.c and args.h

To avoid confusion with variables I changed `verbose` to `VerbosE`. We should find a pattern like this to identify global variables that were set by the arguments.

Also I have added a for loop to look for -- options.
The `argv[i][0] = '\0'` is required  to avoid the string being read by the getopt function.

myzhang1029 commented:
> To avoid confusion with variables I changed `verbose` to `VerbosE`. We should find a pattern like this to identify global variables that were set by the arguments.

This convention is kind of unusual, but we can keep it for now, I'll backport the internal structure in dev-posix2, and such a issue will not exist anymore.

> Also I have added a for loop to look for -- options.
> The `argv[i][0] = '\0'` is required to avoid the string being read by the getopt function.

There is already [`getopt_long`](https://www.gnu.org/software/libc/manual/html_node/Getopt-Long-Options.html) available for parsing -- options. They come with partial matching and other useful features, which will take too long for us to implement one as stable. However, they are GNU extension, although included in BSD and macOS. To make them work on other platforms, we can include one in libpsh, use [gnulib](https://www.gnu.org/software/gnulib) (I personally found `gnulib-tool` frustrating), or use the one included in [myzhang1029/slib](https://github.com/myzhang1029/slib). I somewhat don't like to keep it in the backend, because `getopt_long` isn't defined by POSIX, and such a function is platform-independent after all.

Those solutions all have their drawbacks, if we include one, we'll have to rename them to deal with name-collisions. If we use `gnulib`, our development flow will be longer, and the user compiling would have to clone another git repo. If we use `slib`, another dependency is introduced, although it's my project and it is platform-independent.

I personally prefer including one, but I would like to hear your opinion on this.

myzhang1029 committed:
Remove usage of short and add Doxygen directives

Signed-off-by: Zhang Maiyun <myzhang1029@hotmail.com>

DevManu-de commented:
Im always open for suggestions to improve the code.

> There is already getopt_long available for parsing -- options. They come with partial matching and other useful features, which will take too long for us to implement one as stable. However, they are GNU extension, although included in BSD and macOS. To make them work on other platforms, we can include one in libpsh, use gnulib (I personally found gnulib-tool frustrating), or use the one included in myzhang1029/slib. I somewhat don't like to keep it in the backend, because getopt_long isn't defined by POSIX, and such a function is platform-independent after all.

> Those solutions all have their drawbacks, if we include one, we'll have to rename them to deal with name-collisions. If we use gnulib, our development flow will be longer, and the user compiling would have to clone another git repo. If we use slib, another dependency is introduced, although it's my project and it is platform-independent.

From your previous comment I wasnt able to find out what is wrong with my approach to deal with the -- options.

 My optinion on:
- gnulib: As far as I can tell, it seems really big and may increase the size of psh drastically.
- slib: We should keep keep this in mind eventhough Im not fimiliar with the code. This may help us with other parts of psh.

Personally I would go with my approach because its fairly simple to use and doesnt require us to implement a library.
But Im also open for other suggestions.

DevManu-de committed:
Added and implemented --help flag

Signed-off-by: Manuel Bertele <manuel.bertele2003@gmail.com>

DevManu-de committed:
Made minor changes to --help

Signed-off-by: Manuel Bertele <manuel.bertele2003@gmail.com>

myzhang1029 commented:
> From your previous comment I wasnt able to find out what is wrong with my approach to deal with the -- options.

No, nothing wrong. I just think if a user wanted to type `--debug=true` or `--rcfile=~/new.rc`, (which is also common,) this `strcmp`-based approach will not handle it. However, neither `bash` nor `zsh` handles that, so no worries.


> Personally I would go with my approach because its fairly simple to use and doesnt require us to implement a library.
> But Im also open for other suggestions.

Yes, it is.

Edit: There is something wrong. `--` (no arguments) isn't handled correctly. It is supposed to stop argument processing. And something like `ab--cd` is treated as an argument because it contains `*--*`.

myzhang1029 committed:
Fix long option parser

Signed-off-by: Zhang Maiyun <myzhang1029@hotmail.com>
